server/routes/messagesRoutes.js

```diff
@@ -0,0 +1,131 @@
+router.use(function (req, res, next) {
+    try {
+        jwtUser = verifyToken(req.headers.authorization)
+        next();
+    } catch (err) {
+        console.log(err);
+        res.setHeader('Content-Type', 'application/json');
+        res.end(JSON.stringify({ message: 'Unauthorized' }));
+        res.sendStatus(401);
+    }
+});
+
+
+// Get messages from conversation
+// based on to & from
+router.get('/conversations/query', (req, res) => {
+    let user1 = mongoose.Types.ObjectId(jwtUser.id);
+    let user2 = mongoose.Types.ObjectId(req.query.userId);
+    Message.aggregate([
+        {
+            $lookup: {
+                from: 'users',
+                localField: 'to',
+                foreignField: '_id',
+                as: 'toObj',
+            },
+        },
+        {
+            $lookup: {
+                from: 'users',
+                localField: 'from',
+                foreignField: '_id',
+                as: 'fromObj',
+            },
+        },
+    ])
+        .match({
+            $or: [
+                { $and: [{ to: user1 }, { from: user2 }] },
+                { $and: [{ to: user2 }, { from: user1 }] },
+            ],
+        })
+        .project({
+            'toObj.password': 0,
+            'toObj.__v': 0,
+            'toObj.date': 0,
+            'fromObj.password': 0,
+            'fromObj.__v': 0,
+            'fromObj.date': 0,
+        })
+        .exec((err, messages) => {
+            if (err) {
+                console.log(err);
+                res.setHeader('Content-Type', 'application/json');
+                res.end(JSON.stringify({ message: 'Failure' }));
+                res.sendStatus(500);
+            } else {
+                res.send(messages);
+            }
+        });
+});
+
+// Post private message
+router.post('/', (req, res) => {
+    let from = mongoose.Types.ObjectId(jwtUser.id);
+    let to = mongoose.Types.ObjectId(req.body.to);
+
+    Conversation.findOneAndUpdate(
+        {
+            recipients: {
+                $all: [
+                    { $elemMatch: { $eq: from } },
+                    { $elemMatch: { $eq: to } },
+                ],
+            },
+        },
+        {
+            recipients: [jwtUser.id, req.body.to],
+            lastMessage: req.body.body,
+            date: Date.now(),
+        },
+        { upsert: true, new: true, setDefaultsOnInsert: true },
+        function (err, conversation) {
+            if (err) {
+                console.log(err);
+                res.setHeader('Content-Type', 'application/json');
+                res.end(JSON.stringify({ message: 'Failure' }));
+                res.sendStatus(500);
+            } else {
+                let message = new Message({
+                    conversation: conversation._id,
+                    to: req.body.to,
+                    from: jwtUser.id,
+                    body: req.body.body,
+                });
+
+                req.io.sockets.emit('messages', req.body.body);
```

Comment:
It's not good practice to emit this to all users, instead you should use the socket ids to emit this to just the users in the group

Reply:
Fixed